File: python/generate.py
import os


def generate_c_code(ast):
    int_declarations = set()
    malloc_statements = set()
    code = "#include <stdlib.h>\n#include <string.h>\n\nint main() {"

    for node in ast:
        if node['type'] == 'VariableDeclaration':
            if node['datatype'] == 'int':
                if node['value'] not in int_declarations:
                    code += f"int {node['value']};"
                    int_declarations.add(node['value'])
            elif node['datatype'] == 'str':
                if node['value'] not in malloc_statements:
                    code += f"char *{node['value']} = malloc(256);"
                    malloc_statements.add(node['value'])
        elif node['type'] == 'VariableAssignment':
            if isinstance(node['value'], str):
                code += f"strcpy({node['name']}, {node['value']});"
            else:
                code += f"{node['name']} = {node['value']};"

    for variable in malloc_statements:
        code += f"free({variable});"

    code += "return 0;}"
    return code


def write_to_file(filename, code):
    with open(filename, 'w') as cfile:
        cfile.write(code)


def compile_c_file():
    if os.name != 'nt':
        os.system("cc output.c")


def gencode(ast):
    code = generate_c_code(ast)
    write_to_file('output.c', code)
    compile_c_file()

File: c++/generate.cpp
#include "objects.cpp"
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <vector>
#include <algorithm>

void gencode(std::vector<std::map<std::string, std::string>> &ast) {
    std::vector<std::string> intdecs;
    std::vector<std::string> mallocs;

    std::string code =
        "#include <stdlib.h>\n#include <string.h>\n\nint main() {";

    for (auto i : ast) {
        if (i["type"] == "VariableDeclaration") {

            if (i["datatype"] == "int" &&
                std::find(intdecs.begin(), intdecs.end(), i["value"]) ==
                    intdecs.end()) {
                code += ("int " + i["value"] + " = 0;");
                intdecs.push_back(i["value"]);
                continue;
            }

            if (i["datatype"] == "str" &&
                std::find(mallocs.begin(), mallocs.end(), i["value"]) ==
                    mallocs.end()) {
                code += ("char *" + i["value"] + " = (char*)malloc(256);");
                mallocs.push_back(i["value"]);
                continue;
            }

            continue;
        }

        if (i["type"] == "VariableAssignment") {
            if (i["datatype"] == "str") {
                code += ("strcpy(" + i["name"] + ", " + i["value"] + ");");
                continue;
            } else {
                code += (i["name"] + " = " + i["value"] + ";");
                continue;
            }
        }

        if (i["type"] == "Operation") {
            code += (i["v1"] + " = " + i["v2"] + " " + i["operator"] + " " + i["v3"] + ";");
        }
    }

    for (auto i : mallocs) {
        code += ("free(" + i + ");");
    }
    code += "return 0;}";

    std::ofstream cfile("output.c");
    if (cfile.is_open()) {
        cfile << code;
    }
    cfile.close();
}